### ISSUE TITLE
Bad handling of file encoding for javascript files included in Bridge projects

### ISSUE THREAD

**Issue by DanTup** [bridgedotnet/Bridge#2890]
We discovered a strange issue where we had some source files (CKEditor) that contained code like this:

```js
var something = "\u200b";
```

This is a zero width space. At some point (we suspect minificantion) these appear to be replaced with literal bytes for a zero-width space (this makes sense). However, something goes wrong with the encoding and the character is then replaced with a question mark `?` in the Bridge output.

While testing, we noticed that very strange things happen with file encodings, for example:

- Create two bridge projects, one that references the other
- Add the following content to two javascript files, saved with UTF8 and UTF16 encodings:
    ```js
    var zeroWidthSpaceEscapedUtf8File = "\u200b";
    var zeroWidthSpaceUtf8File = "​";
    ```
- Add this to bridge.json:
    ```js
    "resources": [
      {
        "name": "all.js",
        "files": [
          "utf8.js",
          "utf16.js"
        ]
      }
    ]
    ```
- Reference this project from the second project and then build
- Review the contents of `all.js` in the output folders

### all.js

```text
var zeroWidthSpaceEscapedUtf8File = "\u200b";
var zeroWidthSpaceUtf8File = "​";
v a r   z e r o W i d t h S p a c e E s c a p e d U t f 1 6 F i l e   =   " \ u 2 0 0 b " ; 
 
 v a r   z e r o W i d t h S p a c e U t f 1 6 F i l e   =   "  " ; 
```

### all.min.js

```text
var zeroWidthSpaceEscapedUtf8File="​",zeroWidthSpaceUtf8File="​";v;a;r;z;e;r;o;W;i;d;t;h;S;p;a;c;e;E;s;c;a;p;e;d;U;t;f;F;i;l;e=" \ u 2 0 0 b ";v;a;r;z;e;r;o;W;i;d;t;h;S;p;a;c;e;U;t;f;F;i;l;e="  ";
```

It seems that the encoding of the first file is used for all files being combined, so in this case the utf16 file (which has two-bytes-per-char) is being read as if it was utf8 (resulted in all the extra bytes).

I did also reproduce the issue with the zero-width space being replaced with a literal `?` while typing this up, but I somehow can't repro again now. I think there was something that resulted in an ascii output (I can't repro this now, even with an ascii file thrown in). However I expect fixing this issue would resolve that too.

**Comment by geoffreymcgill** [comment]
Hi @DanTup – Thanks for the excellent report. We are investigating this issue right now and will figure out what changed, then fix.

**Comment by LeonidVeriga** [comment]
@DanTup, 
The default encoding in Bridge is UTF8 (no BOM). When combining resources, Bridge reads the **files** items specified in **resources** config as bytes and puts into `byte[]` buffer as is (the only check performed on is removing BOM sequence if found).

To support the case reported we would need to implement encoding recognition and converting to a resource common one.

**Comment by DanTup** [comment]
@LeonidVeriga I suspected that was the case; but I don't think it's practical (or a great idea) to require the users to ensure that all of their files are always the same encoding. In this case we're pulling a large number of third parties JS files from CKEditor to package into a component. While we could re-encode all of these, we'd have to do it on every update and also ensure that just generally everyone is using the same encoding on all files.

How are you currently reading the files? The [`StreamReader` docs](https://msdn.microsoft.com/en-us/library/7bc2hwcb(v=vs.110).aspx) suggest it's able to set the encoding itself based on the byte order marks?

**Comment by LeonidVeriga** [comment]
@DanTup, yes, I agree (almost absolutely) that it gets some inconvenience. However the current approach works very well with binary resources (fonts, images) - it just reads files as is, embeds and outputs as is.

We will look for implementation options to address your case. Most likely it will be your suggestion (**StreamReader** constructor checking BOM) being applied for **text** resources defined by its file name extention with the following file types considered as text: js, css, ts, html, htm, json, txt, xml.

**Comment by DanTup** [comment]
Does it ever make sense to concatenate binary resources? If not, you could just always do it when concatenating and not have to worry about checking extensions. For single files if you're not writing any BOM and then copying files the resulting file will still be in the original encoding (since you copied the BOM and bytes from the source file) and unchanged if binary.

**Comment by LeonidVeriga** [comment]
No for binary resources. However a single file resource maybe either binary or a text (we do not check resource type by file extention).

The changes addressed in PR #2890:
- If resource has more than one file or header/remark then it is read in its original encoding (BOM checked) and being embedded in the default UTF8 encoding
- Reading resource header with BOM checking
- If resource consists of one file and does not contain any header or remark then it is embedded in its original encoding (old behaviour);

**Comment by DanTup** [comment]
sgtm! Thanks! 👍
